src/main/scala/org/constellation/domain/redownload/RedownloadService.scala

```diff
@@ -1,38 +1,130 @@
+  }
 
-  // TODO: Check for the alignment ownSnapshots <-> majoritySnapshot and call redownload if needed
-  def checkForAlignmentWithMajoritySnapshot(): F[Unit] = ???
+  def recalculateMajoritySnapshot(): F[(Seq[RecentSnapshot], Set[Id])] =
+    for {
+      peerProps <- peersProposals.get
+      ownProps <- ownSnapshots.get
+      ownPropsNormalized = ownProps.mapValues(snap => List((dao.id, Seq(snap))))
+      allProposals = peerProps.mapValues(_.toList) |+| ownPropsNormalized
+      allProposalsNormalized = allProposals.values.flatten.toList
+      peers <- LiftIO[F].liftIO(dao.readyPeers) //todo need testing around join leave, this could cause divergence
+      majority = MajorityStateChooser.chooseMajorWinner(peers.keys.toSeq, allProposalsNormalized)
+      groupedProposals = allProposalsNormalized
+        .groupBy(_._1)
+        .map { case (id, tupList) => (id, tupList.flatMap(_._2)) }
+        .toList
+      snapsThroughMaj = majority.map(maj => MajorityStateChooser.getAllSnapsUntilMaj(maj._2, groupedProposals))
+      majNodeIds = majority.flatMap(maj => MajorityStateChooser.chooseMajNodeIds(maj._2, allProposalsNormalized))
+    } yield (snapsThroughMaj.map(_.sortBy(-_.height)).getOrElse(Seq()), majNodeIds.map(_.toSet).getOrElse(Set()))
+
+  def checkForAlignmentWithMajoritySnapshot(): F[Option[List[RecentSnapshot]]] =
+    for {
+      peers <- LiftIO[F].liftIO(dao.readyPeers)
+      majSnapsIds <- recalculateMajoritySnapshot()
+      ownSnaps <- ownSnapshots.get
+      diff = HealthChecker.compareSnapshotState(majSnapsIds, ownSnaps.values.toList)
+      shouldRedownload = HealthChecker.shouldReDownload(ownSnaps.values.toList, diff)
+      result <- if (shouldRedownload) {
+        logger.info(
+          s"[${dao.id.short}] Re-download process with : \n" +
+            s"Snapshot to download : ${diff.snapshotsToDownload.map(a => (a.height, a.hash))} \n" +
+            s"Snapshot to delete : ${diff.snapshotsToDelete.map(a => (a.height, a.hash))} \n" +
+            s"From peers : ${diff.peers} \n" +
+            s"Own snapshots : ${ownSnaps.values.map(a => (a.height, a.hash))} \n" +
+            s"Major state : $majSnapsIds"
+        ) >>
+          LiftIO[F]
+            .liftIO(dao.healthChecker.startReDownload(diff, peers.filterKeys(diff.peers.contains)))
+            .flatMap(_ => Sync[F].delay[Option[List[RecentSnapshot]]](Some(majSnapsIds._1.toList)))
+      } else {
+        Sync[F].pure[Option[List[RecentSnapshot]]](None)
+      }
+    } yield result
 
 }
 
 object RedownloadService {
-  def apply[F[_]: Concurrent](): RedownloadService[F] = new RedownloadService[F]()
+
+  def deSerProps(nodeSnap: (Id, Array[Array[Byte]])): (Id, Seq[RecentSnapshot]) = nodeSnap match {
+    case (id, serializedProposalMap) =>
+      val proposals: Seq[RecentSnapshot] = serializedProposalMap.flatMap { chunk =>
+        EdgeProcessor
+          .chunkDeSerialize[Seq[(Long, RecentSnapshot)]](chunk, SnapshotInfoChunk.SNAPSHOT_FETCH_PROPOSALS.name)
```

Comment:
Why are you using `SnapshotInfoChunk` to deserialize `own proposals`. It also looks like mixed responsibility.
What's more, `fetch proposals` sounds like a command, not a name of property.